### PR TITLE
[FLINK-19720][table-api] Introduce new Providers and parallelism API

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Introduce new Providers and parallelism API

PR for All Public API.

## Brief change log

- Introduce SourceProvider
- Introduce DataStreamSourceProvider
- Introduce DataStreamSinkProvider
- Introduce ParallelismProvider
- Introduce SINK_PARALLELISM option

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes

### CODE REVIEW

**Review by wuchong — COMMENTED** [review]
Thanks for the work. I left some comments. 

Would be great if @twalthr can help to review the public API.

**Comment by godfreyhe** [comment]
LGTM

**Comment by twalthr** [comment]
Thanks for working on this issue @JingsongLi. I pushed a commit with a couple of suggestions. Feel free to undo whatever you don't like.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/connector/sink/DataStreamSinkProvider.java [hunk 1]
```diff
@@ -24,16 +24,16 @@
 import org.apache.flink.table.data.RowData;
 
 /**
- * Provider to consume a {@link DataStream} as a runtime implementation for
- * {@link DynamicTableSink}.
+ * Provider that consumes a Java {@link DataStream} as a runtime implementation for {@link DynamicTableSink}.
  *
- * <p>NOTE: This is only for advanced connector developers.
+ * <p>Note: This provider is only meant for advanced connector developers. Usually, a sink should consist
+ * of a single entity expressed via {@link OutputFormatProvider} or {@link SinkFunctionProvider}.
```
Comment (wuchong) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/connector/sink/DataStreamSinkProvider.java:30]: +1 for this. This makes the responsity of this interface more clear.

**Comment by JingsongLi** [comment]
Thanks @wuchong @godfreyhe @twalthr for your review, I'll merge it.

**Pull request merged**